### ISSUE TITLE
Implement receiveOrClosed for channels and select clause

### ISSUE THREAD

**Issue by qwwdfsad** [Kotlin/kotlinx.coroutines#330]
We need `receiveOrClosed` method on top of channels and `onReceiveOrClosed` for `select` which will have return type `Either<E, CloseToken>`.

Currently, pattern to work with channels always has the following form which is clumsy and obfuscates code:
```
 try {
    source.onReceive { ... }
} catch (e: ClosedReceiveChannelException) {
  // source is closed
}
```

The better alternative may be 
```
source.onReceiveOrClosed {
  when (it) {
    CloseToken -> // end up consumption
    _ -> // proceed normally
  }
}
```

For implementation to be efficient we need inline classes (https://github.com/Kotlin/KEEP/issues/104)

**Comment by fvasco** [comment]
I agree with you, but in my experience the `onClose` is missing, this can become `channel.job.onJoin`

**Comment by elizarov** [comment]
Let me clarify that the considerable fraction of value is going to come from support for `onReceiveOrClosed` clause in `select` expression. With that you can write operations on multiple channels with less boilerplate, without having to deal with concurrency, and with ability to code any behavior on the close of any of the source channels. 

Note, that when we have `channel.job.onJoin` you still will NOT be able to use `onJoin` in `select` together with `onReceive`, e.g this:

```
select {
    channel.onReceive { ... }
    channel.job.onClose { ... }
}
```

will not work as expected. We shall make sure it fails fast in this case with a clear error message.

Another note, is that we should not actually expose `CloseToken` in public API, since trying to send this token over another channel will lead to unpredictable results, so we shall encapsulate it by introducing a special `ValueOrClosed<T>` type (that will be implement as `inline class` in the future).

**Comment by fvasco** [comment]
You are right @elizarov
however I still prefer something like:

```kotlin
select {
    channel.onClose { ... } // onClose first
    channel.onReceive { ... }
}
```

Kotlin/KEEP#104 is WIP, but I cannot understand how is it possible apply to `onReceiveOrClosed`, the only way to implement a `ValueOrClosed<T>` is a sealed class (and sealed class cannot be inlined).

**Comment by elizarov** [comment]
See also this SO discussion https://stackoverflow.com/questions/51463941/how-to-safely-select-across-channels-where-some-may-get-concurrently-closed/

**Comment by qwwdfsad** [comment]
Will be named `receiveCatching` and `onReceiveCatching`, see https://github.com/Kotlin/kotlinx.coroutines/issues/974#issuecomment-806569582

**Comment by matejdro** [comment]
`receiveOrClosed` implementation mentions that it will not be internal once https://youtrack.jetbrains.com/issue/KT-27524 is fixed.

That bug has been fixed. What is the status for this?

**Comment by qwwdfsad** [comment]
It will be implemented in coroutines 1.5.0 under `receiveCatching` and `onReceiveCatching` names
